oak_restricted_kernel/src/ghcb.rs

```diff
@@ -0,0 +1,187 @@
+    registers::control::Cr3,
+    structures::paging::{
+        page_table::{PageTable, PageTableFlags},
+        PageSize, Size1GiB, Size2MiB, Size4KiB,
+    },
+};
+
+/// The mask for the encrypted bit
+const ENCRYPTED_BIT: u64 = 1 << ENCRYPTED_BIT_POSITION;
+
+/// A wrapper to ensure that the GHCB is alone in a 2MiB page.
+///
+/// We use 2MiB pages during early boot, so we must make sure there are no other kernel data
+/// structures located in the page so that we can safely share the page with the hypervisor without
+/// leaking other data.
+#[derive(Debug)]
+#[repr(C, align(2097152))]
+struct GhcbAlignmentWrapper {
+    ghcb: Ghcb,
+}
+
+static_assertions::assert_eq_size!(GhcbAlignmentWrapper, [u8; Size2MiB::SIZE as usize]);
+
+static mut GHCB_WRAPPER: GhcbAlignmentWrapper = GhcbAlignmentWrapper { ghcb: Ghcb::new() };
+
+/// Initializes the GHCB and shares it with the hypervisor during early boot.
+pub fn init_ghcb_early(snp_enabled: bool) -> GhcbProtocol<'static, Ghcb> {
+    // Safety: This is called only during early boot, so there is only a single execution context.
+    let ghcb = unsafe { &mut GHCB_WRAPPER.ghcb };
+    let translate = |vaddr: VirtAddr| {
+        let address = vaddr.as_u64();
+        // We assume a fixed offset mapping for kernel pages.
+        if address < KERNEL_OFFSET {
+            None
+        } else {
+            Some(PhysAddr::new(address - KERNEL_OFFSET))
+        }
+    };
+    share_ghcb_with_hypervisor(ghcb, snp_enabled, translate);
+    ghcb.reset();
+
+    GhcbProtocol::new(ghcb, translate)
+}
+
+/// Marks the page containing the GHCB data structure to be shared with the hypervisor.
+fn share_ghcb_with_hypervisor<VP: Translator>(ghcb: &Ghcb, snp_enabled: bool, translate: VP) {
+    let ghcb_address = VirtAddr::new(ghcb as *const Ghcb as usize as u64);
+    // On SEV-SNP we need to additionally update the RMP and register the GHCB location with the
+    // hypervisor.
+    if snp_enabled {
+        // It is OK to crash if we cannot find a valid physical address for the GHCB.
+        let ghcb_physical_address = translate(ghcb_address).unwrap();
+        mark_2mib_page_shared_in_rmp(ghcb_physical_address);
+        let ghcb_location_request =
+            RegisterGhcbGpaRequest::new(ghcb_physical_address.as_u64() as usize)
+                .expect("Invalid address for GHCB location");
+        register_ghcb_location(ghcb_location_request)
+            .expect("Couldn't register the GHCB address with the hypervisor");
+    }
+    // Mark the page as not encrypted in the page tables.
+    set_encrypted_bit_for_page(&ghcb_address, false);
+}
+
+/// Sets the encrypted bit for the page that contains the address.
+///
+/// Since this is called during early boot, before the frame allocator or page mapper is initialised
+/// we have to manually traverse the existing page tables. We use only 2MiB huge pages during early
+/// boot.
+fn set_encrypted_bit_for_page(address: &VirtAddr, encrypted: bool) {
```

Comment:
Ah, good old manual page walk :)

I think you can use `EncryptedPageTable` from `mm/encrypted_mapper.rs` here. It has support for changing the encrypted bit of a page already.